### PR TITLE
Fix issue #587.

### PR DESCRIPTION
**Pull request by guzman-raphael**
Will convert array from python object to appropriate data type if all elements are of the same type.

### CODE REVIEW

File: datajoint/fetch.py [hunk 1]
```diff
@@ -172,6 +172,12 @@ def __call__(self, *attrs, offset=None, limit=None, order_by=None, format=None,
                     list((to_dicts if as_dict else lambda x: x)(ret[self._expression.primary_key])) if is_key(attribute)
                     else ret[attribute] for attribute in attrs]
                 ret = return_values[0] if len(attrs) == 1 else return_values
+                
+                if isinstance(ret,(np.ndarray)) and len(ret) > 0:
+                    main_type = type(ret[0])
+                    print(main_type)
```
Comment (dimitri-yatsenko) [inline datajoint/fetch.py]: remove debug code

File: datajoint/fetch.py [hunk 2]
```diff
@@ -182,6 +188,11 @@ def __call__(self, *attrs, offset=None, limit=None, order_by=None, format=None,
                 ret = np.array(ret, dtype=heading.as_dtype)
                 for name in heading:
                     ret[name] = list(map(partial(get, heading[name]), ret[name]))
+
+                    if isinstance(ret[name],(np.ndarray)) and len(ret[name]) > 0:
+                        main_type = type(ret[name][0])
+                        if all(isinstance(i, main_type) for i in ret[name]) and not isinstance(ret[name][0],str):
```
Comment (dimitri-yatsenko) [inline datajoint/fetch.py]: could check for `str` before all the other checks for better efficiency.

File: tests/test_fetch_same.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            {'id' : 1, 'resp' : 94.3,'sim' : 34.23},
+            {'id' : 2, 'resp' : 1.90,'sim' : 10.23}    
+        ]
+
+        ProjData().insert(data_ins)
+
+        trials_new = ProjData.proj(new='resp-sim')
+        new = trials_new.fetch('new')
+
+        assert_equal(new.dtype,np.float64)
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: PEP8

File: datajoint/fetch.py [hunk 4]
```diff
@@ -182,6 +188,11 @@ def __call__(self, *attrs, offset=None, limit=None, order_by=None, format=None,
                 ret = np.array(ret, dtype=heading.as_dtype)
                 for name in heading:
                     ret[name] = list(map(partial(get, heading[name]), ret[name]))
+
+                    if isinstance(ret[name],(np.ndarray)) and len(ret[name]) > 0:
```
Comment (dimitri-yatsenko) [inline datajoint/fetch.py]: use a more efficient way to test whether `ret` is empty

File: datajoint/fetch.py [hunk 5]
```diff
@@ -172,6 +172,12 @@ def __call__(self, *attrs, offset=None, limit=None, order_by=None, format=None,
                     list((to_dicts if as_dict else lambda x: x)(ret[self._expression.primary_key])) if is_key(attribute)
                     else ret[attribute] for attribute in attrs]
                 ret = return_values[0] if len(attrs) == 1 else return_values
+                
+                if isinstance(ret,(np.ndarray)) and len(ret) > 0:
```
Comment (dimitri-yatsenko) [inline datajoint/fetch.py]:
```suggestion
                if isinstance(ret, np.ndarray) and ret:
```

File: datajoint/fetch.py (hunk 1 above)
Comment (guzman-raphael) [reply datajoint/fetch.py]: thank you. I have made sure these have been removed.

File: datajoint/fetch.py (hunk 2 above)
Comment (guzman-raphael) [reply datajoint/fetch.py]: I agree. I have now moved the operation to prior to initializing the np.array.

File: tests/test_fetch_same.py (hunk 3 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: ty! Updated.

File: datajoint/fetch.py (hunk 4 above)
Comment (guzman-raphael) [reply datajoint/fetch.py]: Addressed.

File: datajoint/fetch.py (hunk 5 above)
Comment (guzman-raphael) [reply datajoint/fetch.py]: thanks

File: tests/test_fetch_same.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
+from nose.tools import assert_true, assert_false, assert_equal, \
+                        assert_list_equal, raises
+from . import PREFIX, CONN_INFO
+import datajoint as dj
+# import importlib
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: removed commented code

File: tests/test_fetch_same.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# dj = importlib.import_module('datajoint-python.datajoint', None)
+import numpy as np
+
+Schema = dj.schema(PREFIX + '_fetch_same', connection=dj.conn(**CONN_INFO))
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: Our conventions is to keep the `schema` object in lower case. For example `dj.Diagram` will look for the `schema` object in the module by that name.

File: tests/test_fetch_same.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    definition = """
+    id : int
+    ---
+    resp : float
+    sim  : float
+    big : longblob
+    blah : varchar(10)
+    """
+
+data_ins = [
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: use complete English words for variable names.

File: tests/test_fetch_same.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    {'id': 0, 'resp': 20.33, 'sim': 45.324, 'big': np.random.randn(10, 5), 'blah': 'yes'},
+    {'id': 1, 'resp': 94.3, 'sim': 34.23, 'big': {'key1': np.random.randn(20, 10)}, 'blah': 'si'},
+    {'id': 2, 'resp': 1.90, 'sim': 10.23, 'big': np.random.randn(4, 2), 'blah': 'sim'}
+]
+
+ProjData().insert(data_ins)
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: Since `data_ins` is only used once, then there is no good reason to assign it to a variable.  Just insert the literal object directly.

File: tests/test_fetch_same.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class TestFetchSame:
+
+    @staticmethod
+    def test_object_conversion_one():
+
+        trials_new = ProjData.proj(sub='resp-sim')
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: There is no reason to create the `trials_new` variable. You can fetch from the query expression directly.

File: tests/test_fetch_same.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        new = trials_new.fetch('sub')
+
+        assert_equal(new.dtype, np.float64)
+
+    @staticmethod
+    def test_object_conversion_two():
+
+        trials_new = ProjData.proj(sub='resp-sim', add='resp+sim')
+        new = trials_new.fetch('sub', 'add')
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: assign `sub` and `add` to different variables for clarity

File: tests/test_fetch_same.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+        assert_equal(new[0].dtype, np.float64)
+        assert_equal(new[1].dtype, np.float64)
+
+    @staticmethod
+    def test_object_conversion_all():
+
+        trials_new = ProjData.proj(sub='resp-sim', add='resp+sim')
```
Comment (dimitri-yatsenko) [inline tests/test_fetch_same.py]: there is no reason for creating  the `trials_new` variable. You can fetch directly from the query expression.

File: tests/test_fetch_same.py (hunk 6 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: This is related to my debug setup. I have updated it to properly determine which import strategy is best.

File: tests/test_fetch_same.py (hunk 7 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: Updated.

File: tests/test_fetch_same.py (hunk 8 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: OK

File: tests/test_fetch_same.py (hunk 9 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: Updated.

File: tests/test_fetch_same.py (hunk 10 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: Updated.

File: tests/test_fetch_same.py (hunk 11 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: Updated.

File: tests/test_fetch_same.py (hunk 12 above)
Comment (guzman-raphael) [reply tests/test_fetch_same.py]: Updated.
